packages/ai-chat/src/common/chat-model.ts

```diff
@@ -0,0 +1,753 @@
+    readonly requestId: string;
+    readonly progressMessages: ChatProgressMessage[];
+    addProgressMessage(message: { content: string } & Partial<Omit<ChatProgressMessage, 'kind'>>): ChatProgressMessage;
+    getProgressMessage(id: string): ChatProgressMessage | undefined;
+    updateProgressMessage(message: { id: string } & Partial<Omit<ChatProgressMessage, 'kind'>>): void;
+    readonly response: ChatResponse;
+    readonly isComplete: boolean;
+    readonly isCanceled: boolean;
+    readonly isError: boolean;
+    readonly agentId?: string
+    cancel(): void;
+    error(error: Error): void;
+    readonly errorObject?: Error;
+
+}
+
+/**********************
+ * Implementations
+ **********************/
+
+export class ChatModelImpl implements ChatModel {
+    protected readonly _onDidChangeEmitter = new Emitter<ChatChangeEvent>();
+    onDidChange: Event<ChatChangeEvent> = this._onDidChangeEmitter.event;
+
+    protected _requests: ChatRequestModelImpl[];
+    protected _id: string;
+
+    constructor(public readonly location = ChatAgentLocation.Panel) {
+        // TODO accept serialized data as a parameter to restore a previously saved ChatModel
+        this._requests = [];
+        this._id = generateUuid();
+    }
+
+    getRequests(): ChatRequestModelImpl[] {
+        return this._requests;
+    }
+
+    get id(): string {
+        return this._id;
+    }
+
+    addRequest(parsedChatRequest: ParsedChatRequest, agentId?: string): ChatRequestModelImpl {
+        const requestModel = new ChatRequestModelImpl(this, parsedChatRequest, agentId);
+        this._requests.push(requestModel);
+        this._onDidChangeEmitter.fire({
+            kind: 'addRequest',
+            request: requestModel,
+        });
+        return requestModel;
+    }
+
+    isEmpty(): boolean {
+        return this._requests.length === 0;
+    }
+}
+
+export class ChatRequestModelImpl implements ChatRequestModel {
+    protected _id: string;
+    protected _session: ChatModel;
+    protected _request: ChatRequest;
+    protected _response: ChatResponseModelImpl;
+    protected _agentId?: string;
+
+    constructor(session: ChatModel, public readonly message: ParsedChatRequest, agentId?: string) {
+        // TODO accept serialized data as a parameter to restore a previously saved ChatRequestModel
+        this._request = message.request;
+        this._id = generateUuid();
+        this._session = session;
+        this._response = new ChatResponseModelImpl(this._id, agentId);
+        this._agentId = agentId;
+    }
+
+    get id(): string {
+        return this._id;
+    }
+
+    get session(): ChatModel {
+        return this._session;
+    }
+
+    get request(): ChatRequest {
+        return this._request;
+    }
+
+    get response(): ChatResponseModelImpl {
+        return this._response;
+    }
+
+    get agentId(): string | undefined {
+        return this._agentId;
+    }
+}
+
+export class ErrorResponseContentImpl implements ErrorResponseContent {
```

Comment:
These classes are meant to prevent developer errors. There will be many agents wanting to represent an error content, string content etc. If they would like the default behavior they can just reuse these impls without reimplementing `asString`, `merge`, `kind` in the same way as the base framework.

Reply:
There really is no interesting code in this class. Do we really expect folks to screw up this code?
``` 
catch (e) {
   return { kind: ErrorResponse.KIND', e };
}
```

Reply:
Well that code is already wrong as it will not expose an `error` field but an `e` field. Also depending on the outcome of the `asString` discussion it also does not offer a `asString` method.

Reply:
Users can _still_ do this even if we provide the implementation class. It sometimes makes sense that we provide default implementations of behavior when it's interesting, but this is so trivial, it is just adding noise 🤷